traffic_ops/traffic_ops_golang/plugin/hello_world.go

```diff
@@ -0,0 +1,34 @@
+package plugin
+
+/*
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import (
+	"strings"
+)
+
+func init() {
+	AddPlugin(10000, Funcs{onRequest: hello})
+}
+
+const HelloPath = "/_hello"
+
+func hello(d OnRequestData) IsRequestHandled {
+	if !strings.HasPrefix(d.R.URL.Path, HelloPath) {
```

Comment:
I know this feels a little odd, rather than adding a path to serve as part of the registration. But with Grove, we found it was a lot more flexible to have a generic OnRequest that can do anything, and instruct the main app to stop or continue processing. The idiom with Grove is to prefix path plugins with `http_`, so e.g. `plugins/http_routing_override.go`, which IMO would be a good idiom to continue in TO.

Reply:
Was this discussed on the Dev List, because I think there are some valid arguments for a Microservice approach as well.

Reply:
It'd be very easy to create a plugin which calls an external service, or a set of plugins with shared config as complex as you like. A specific "microservice extensions" framework would be strictly less powerful than this. 

>Was this discussed on the Dev List?

It wasn't, this is just reproducing the existing Perl TO functionality, but feel free to.

Reply:
+1 to what @rob05c says. This framework can be used to intercept any request, and how the plugin developer chooses to deal with the intercepted request is up to them. This could be a plugin integrated directly with the rest of Golang TO, or it could be a call out to a separate microservice outside of TO.

Conceptually this appears to be a 1:1 implementation of the basic functionality that we have in Perl/Mojolicious today for extensions. In the Mojo case, you could also intercept a route and within the implementing code, could call out to a microservice or could implement Perl code. That's how this code works too, the only difference is that we're now in a compiled vs interpreted environment.

Reply:
I guess my main concern is now we have to build a more sophisticated build system to manage the rpm versions because the binary would be different based upon the plugins that are "hooked" in.  For example, how could I look at the rpm and determine it had "my custom plugins" in it easily?  The githash and TC version wouldn't be enough imo.

I also don't understand how this plugin (compiled in) approach would be more powerful than a traditional MS approach?

Reply:
You are correct that the last version numbers could be different across TC if you're supplying custom plugins.  That's a good thing though in my mind.  Traditionally you always assemble your repository state then build the result.  TC has currently embraced one repo and one application version to rule all components; a fairly normal practice among many with its pros and cons. The githash is the source of truth for what's included since it's the state the repo was at build time.  If someone had been patching perl after deployment, it's nolonger that version that was built anyway and nolonger has a single commit to reference the code state that was built.  The big bonus for this kind of plugin system is that when/if you're knitting repositories together you're less likely to have conflicts which results in lower maintenance costs and providing a template for writing more modular and therefore testable code.

Comparing a plugin system like this to a microservice architecture in the end achieve similar goals.  Microservices give you scalpel-like ability to adjust what small functions are doing.  In practice you pay a significant price for those gains when having to triage why things aren't working perfectly and in maintaining compatibility matrices between all the components.  The latter case is mitigated if you upgrade everything everytime, but then you're not using the gain you designed for.  A plugin system like this is much simpler to diagnose problems with because you're working with single binaries with any variances locked at build time.

Reply:
>I also don't understand how this plugin (compiled in) approach would be more powerful than a traditional MS approach?

Because it's capable of both.

```
func init() { AddPlugin(10000, Funcs{onRequest: helloMicroService}) }
const HelloServicePath = "/api/1.2/cdns/routing"
const MicroserviceURI = "https://example.net"
func helloMicroService(d OnRequestData) IsRequestHandled {
	if !strings.HasPrefix(d.R.URL.Path, HelloServicePath) {
		return RequestUnhandled
	}
	resp, _ := http.Get(MicroserviceURI)
	w.WriteHeader(resp.StatusCode)
	io.Copy(w, resp.Body)
	return RequestHandled
}
```

So you see, this plugin system can be used for microservices. And since it can also be used for direct code, it's more powerful than a microservice-only extension system.

>now we have to build a more sophisticated build system

Our build system already creates an RPM from an internal merge with our extensions. We won't need to change it at all. And as @jhg03a said, it'll have a unique internal git hash we can use to identify which plugins were put in.

>how could I look at the rpm and determine it had "my custom plugins" in it easily?

That's a really good idea, actually: we should make the traffic_ops build.sh add the list of plugins to the Yum Info in the RPM. I can add that, either to this PR, or a following one.

Reply:
@dewrich Ok, I added the list of plugins to the RPM:

```bash
$ rpm -qip dist/traffic_ops-2.3.0-8898.4abdfa71.el7.x86_64.rpm                                                                                                                                                                                                                                                          
Name        : traffic_ops
⋮
Description :
Traffic Ops is the tool for administration (configuration and monitoring) of all components in a Traffic Control CDN.

This package provides Traffic Ops with the following plugins:
hello_config
hello_context
hello_shared_config
hello_startup
hello_world
```

Does that address all your concerns?

Reply:
Partially, I guess my thought was more around the fact that your `dist/traffic_ops-2.3.0-8898.4abdfa71.el7.x86_64.rpm` has no obvious indication of that plugin list, so now we're still juggling rpm's that have been compiled with the plugins needed.  If I somehow plugin XYZ got missed now I have to go rebuild the rpm